docs/dev/truncate-with-tablets.md

```diff
@@ -0,0 +1,79 @@
+# Background
+
+The CQL TRUNCATE query is obviously data-modifying since it deletes user data, however,
+unlike DELETE queries, it is not based on a tombstone that carries a write-timestamp,
+with which we can determine for each cell, if it was logically written before
+or after the tombstone timestamp.
+
+Instead, truncate propagates in the cluster using an out-of-band RPC and is currently
+applied locally in each node using the local wall clock of each node (not synchronized
+with each other), typically after flushing the tables and taking a snapshot of them,
+if auto-snapshot is enabled.
+
+This presents a problem when tablets are enabled since there could on-going tablet
+transitions in the background, that the user has no control on, like tablet migration,
+with which a tablet might escape truncate by migrating between nodes just at the wrong
+time so it won’t be considered for truncate by any node.
+
+# Introduction
+
+This design proposal relates to [Make TRUNCATE TABLE safe with tablets #16411](https://github.com/scylladb/scylladb/issues/16411).
+
+Originally, the plan was to use a Truncate tombstone as per [Truncate must be based on a tombstone #11230](https://github.com/scylladb/scylladb/issues/11230).
+
+This document lists the currently proposed solutions in order to reach consensus on how
+truncate should be implemented.
+
+# Proposed solutions
+
+## 1. Short term mitigation
+
+We can pause any background load balancing operations (if any are running) and execute the
+truncate as it is currently implemented. We would issue a RPC which performs a memtable flush,
+then take a snapshot of the SSTables if auto-snapshot is enabled, and delete the SSTables
+belonging to the given table and associated views.
+
+## 2. Truncate as a topology operation
+
+We need a barrier that will wait on any on-going tablet transitions on the truncated table
+AND on its subordinate views (materialized views and secondary indexes), and prevent any new
+ones from taking place, similar to safe repair (See [Serialize repair with tablet migration #18641](https://github.com/scylladb/scylladb/pull/18641))
+
+At this point we can build a transition plan, using a new, “truncate” tablet transition type,
+for all tablet replicas for the truncate table and its view.
+
+Open questions: should flush and snapshot happen as they do today, before preparing and executing
+the truncate transition plan, or should they be performed inline, by each tablet, as it is truncated.
+For the latter, we’ll need to pass the snapshot tag (a.k.a name) so the distributed snapshots will
+use the same tag. This is harder, but we want to support a distributed snapshot anyhow in the future
+for integrating backup with tablets (and the object storage stack)
```

Comment:
It's fine to keep the current practice, especially as this is an interim solution. The guideline is to be no worse than vnodes, we don't have to be better for this rare operation.